ftfy/chardata.py

```diff
@@ -54,6 +54,21 @@ def _build_regexes():
 ENCODING_REGEXES = _build_regexes()
 
 
+def _build_html_uppercase_entities():
+    uppercase_entities = {}
+    for name, codept in html.entities.name2codepoint.items():
+        if codept < 0x300:
+            name_upper = name.upper()
+            entity_upper = '&' + name_upper + ';'
+            if name_upper != name and html.unescape(entity_upper) == entity_upper:
```

Comment:
`html.entities.html5` looks good -- I wasn't aware of that one. I'll go switch to that, and remove the redundant check.

Reply:
I wanted to limit the rule to mostly Latin letters, and particularly symbols in Latin-1, because those are the ones where encoding errors tend to slip through. If this was dealing with text in Greek, every character would end up encoded this way. You'd think someone would notice that their text looks like `&kappa;&alpha;&kappa;&omicron;&sigmaf;` before they took the next step of running the all-caps-ifier on it.

If we support too many symbols, the rules get very ambiguous. When we're dealing with letters, `&NTILDE;` could be the capitalization of `&ntilde;` or `&Ntilde;`, but we know to go with the capital one because, hey, they seem to want capital letters here. But for example, `&And;` (`⩓`) is a doubled-up version of `&and;` (`∧`) that I've never seen used.

Even `&Not;` (`⫬`) exists, and I went with `&not;` (`¬`) because it's the one in Latin-1 and the one that might be further decoded as mojibake.

I guess I could have a different rule of supporting all these symbols, and in the ambiguous cases we take the one whose name is in lowercase and then uppercase the result, but then the regex would have to match some of these very long entity names like `&leftrightharpoons;` at the cost of performance.

Reply:
Ah well, I suppose even the ones under U+300 have long names in HTML5 now. Like `&NonBreakingSpace;`.

Looking into the alternate rule of, only use all-lowercase names, and uppercase the result.

Reply:
What, I wonder, are the chances that something ends up in allcaps other than a capitalized letter like NTILDE or SACUTE?  That is, someone who wants a capitalized `ntilde` might type `NTILDE`, or someone trying to capitalize text might call a capitalizing function on an HTML escape; but what would cause someone to type `&NOT;` or call a capitalizing function on something with `&not;` in it?

That is--maybe it's plausible that it would happen, but on the other hand maybe setting up a dictionary of only alphabetic characters would suffice.

Reply:
Oh no.

Did you know that HTML5 can decode the string `this&nbspis&nbspterrible` as `this is terrible`? It seems there's a state machine that's constantly matching entity names, and the common ones don't even require semicolons.

I wouldn't know how to efficiently reproduce that so we can decode `THIS&NBSPIS&NBSPTERRIBLE`.

Reply:
My reason to want to decode `&NOT;` as `¬` is imagining this data flow:

- Person A, who wants to use Unicode but is bad at it, mixes up UTF-8 and Latin-1, creating strings that occasionally have the usual mojibake characters like `Ã¬`.
- Person B, entirely afraid of Unicode, writes a function that encodes all non-ASCII input as HTML escapes.
- Person C, who wants to normalize away capitalization differences, uppercases all the text that's input to their function.

A, B, and C's functions end up as a pipeline, in that order, and someone wants ftfy to untangle it.

Seems a bit farfetched but I have already seen worse things. A happens all the time and we've already seen B and C in a widely-used pipeline.

Reply:
Anyway I want to draw the line somewhere, so one way to do it that's almost like what I did so far is:

- Single characters with codepoints under U+300
- Whose names are all lowercase and end in a semicolon
- Use the uppercase version of the decoded result when that's relevant